keras_nlp/utils/text_generation.py

```diff
@@ -288,63 +291,100 @@ def token_probability_fn(inputs):
         )
+    if num_beams == 1:
+        return greedy_search(
+            token_probability_fn=token_probability_fn,
+            prompt=prompt,
+            max_length=max_length,
+            end_token_id=end_token_id,
+            pad_token_id=pad_token_id,
+        )
 
     prompt = _validate_prompt(prompt)
 
-    if prompt.shape[-1] == 0:
-        raise ValueError(
-            "Length of `prompt` is 0, please provide a non-empty `prompt`."
-        )
-
     input_is_1d = prompt.shape.rank == 1
     if input_is_1d:
         prompt = prompt[tf.newaxis, :]
-    _validate_token_probability_fn(token_probability_fn, prompt)
 
-    batch_size, length = prompt.shape
+    batch_size, length = _get_prompt_shape(prompt)
+    prompt, mask = _pad_prompt(prompt, max_length)
+
+    vocab_size = _validate_token_probability_fn(token_probability_fn, prompt)
+
     if length >= max_length:
         return tf.squeeze(prompt) if input_is_1d else prompt
 
-    # Initialize beam.
-    beams = tf.expand_dims(prompt, 1)
-    beams_prob = tf.zeros([batch_size, 1])
-    i = length
-    while i < max_length:
-        beam_size = beams.shape[1]
-        beam_preds = []
-        for j in range(beam_size):
-            preds = token_probability_fn(beams[:, j, :])
-            if from_logits:
-                preds = keras.activations.softmax(preds, axis=-1)
-            beam_preds.append(preds)
-        stacked_preds = tf.stack(beam_preds, axis=1)
-        vocab_size = stacked_preds.shape[2]
-        logits = tf.reshape(stacked_preds, [batch_size, beam_size * vocab_size])
+    # Initialize beam with shape `(batch_size, num_beams, length)`.
+    beams = tf.repeat(tf.expand_dims(prompt, axis=1), num_beams, axis=1)
+
+    # Initialize `beams_prob`` with shape `(batch_size, num_beams)`
+    beams_prob = tf.zeros([batch_size, 1], dtype=tf.float32)
+    beams_prob = tf.concat(
+        [beams_prob, tf.fill((batch_size, num_beams - 1), -3.40282e38)],
+        axis=-1,
+    )
+
+    def one_step(beams, beams_prob, length):
+        truncated_beams = beams[..., :length]
+
+        # logits is of shape [batch_size, num_beams * vocab_size].
+        flattened_beams = tf.reshape(
+            truncated_beams, shape=[batch_size * num_beams, -1]
+        )
+        logits = token_probability_fn(flattened_beams)
+        logits = tf.reshape(logits, shape=[batch_size, -1])
+
         probs = tf.math.log(logits) + tf.repeat(
             beams_prob, repeats=vocab_size, axis=1
         )
-        num_beams = min(beam_size * vocab_size, num_beams)
+
         candidate_prob, candidate_indexes = tf.math.top_k(
             probs, k=num_beams, sorted=False
         )
         candidate_beam_indexes = candidate_indexes // vocab_size
         next_token = candidate_indexes % vocab_size
 
         beams = tf.gather(beams, candidate_beam_indexes, axis=1, batch_dims=1)
-        beams = tf.concat([beams, next_token[..., tf.newaxis]], axis=-1)
+
+        next_token = tf.where(
```

Comment:
Suggest for this whole block something more like this...

```python
# Build a new column of updates to scatter into the beam tensor.
next_tokens = tf.where(
    condition=mask[..., length, tf.newaxis],
    x=beams[..., length],
    y=next_token,
)
next_tokens = tf.flatten(next_token)

# Generate `(batch_index, beam_index)` tuples for each beam. 
beam_indices = tf.where(tf.ones((batch_size, num_beams), tf.bool))
beam_indices = tf.cast(beam_indices, length.dtype)
# Build a tensor repeated `length` values.
length_indices = tf.fill((batch_size * num_beams, 1), length)
# Concatenate to a triplet of `(batch_index, beam_index, length)`.
indices = tf.concat([beam_indices, length_indices], axis=1)

beams = tf.tensor_scatter_nd_update(
    tensor=beams,
    indices=indices,
    updates=next_tokens,
)
```

Feel free to rename as needed, but overall you should look out for a few things.

 - Break up complex nested lines into intermediate lines that fit our line length when possible. 
 - Use helpful naming for intermediate variables to indicate the flow.
 - Do not add comments everywhere, but add brief comments in places the code itself could use clarification.

The end goal should be making this very complex list of ops someone more readable to a newcomer trying to figure this out.

Reply:
Will keep that in mind, thanks a ton! Looks cleaner now